### DISCUSSION TITLE
Question on Manual Report Generation

### DISCUSSION THREAD

**Discussion by firdaussoberi** [massimo-zaniboni/asterisell-v5#17] [Q&A] [answered]
Hi it's been a while,

I just got a question on report generation, manually through script/job.
I know that currently it is handled by the scheduler and each report will get sent over email, 
but let's say we have some missed reports that we want to generate manually, which is also good in case of manual bulk report generations. 

How would we go to do that?
Basically currently if we want to manually create the report, we have to go to this view and click the 'Generate Report' button at the bottom:

<img width="1280" alt="Screenshot 2022-05-12 at 5 00 13 PM" src="https://user-images.githubusercontent.com/8426299/168035278-a4307d08-a61f-4466-9679-f8333ee5ffbe.png">
<img width="1280" alt="Screenshot 2022-05-12 at 5 09 16 PM" src="https://user-images.githubusercontent.com/8426299/168035310-649212fa-0c79-4da4-bb8b-4a71be662c0b.png">

What if instead we want to bypass the button and achieve the same report, this time through a script,
is there a way to do it and if possible could you guide us? 

Thank you so much.
Best Regards,
Firdaus

**Comment by massimo-zaniboni** [comment, marked as answer]
> What if instead we want to bypass the button and achieve the same report, 
> this time through a script,
> is there a way to do it and if possible could you guide us?

The quick answer in NO, because up to date Asterisell has no API. 

Maybe one can add a command to ``php asterisell.php`` utility, where it is specified the report to (re)generate, and this command can be called from a script. It is easier than adding an API to Asterisell.

**Reply by firdaussoberi** [reply]
I see. That option would be good enough as we are not looking to call the script from outside the program/network. 

I've meddled with the file before to add custom install function linked to ```asterisell_instances.py```.
Do you know where is the part calling the button, I mean what subsequent functions (or single main function if any) are called when the ```Generate Report```  button is invoked, if you could point for us?

Otherwise I think we have to dig further to replicate the button/report generation, bulk of it I tracked to be in 
```apps/asterisell/lib/jobs/reports/invoices/GenerateInvoiceUsingTemplate01PDF.php```
```apps/asterisell/lib/jobs/reports/invoices/InvoiceCalcStore.php```

Thanks

**Comment by massimo-zaniboni** [comment]
The report generation code is rather complex because various customers asked me different features. 

I cannot point to a single point of the code, and say: call this.

**Reply by firdaussoberi** [reply]
I see 😂. So it sort of intertwined in the view itself, I get what you mean. 
We'll try to see what's best first, might need your help again I reckon.

Thanks for the clarification!
